### ISSUE TITLE
InvalidOperationException calling ServiceSerializer.Serialize method

### ISSUE THREAD

**Issue by asierpn** [Azure/azure-mobile-apps#900]
## Describe the bug

We use the DatasyncClient.Serializer in our custom implementation of the old InvokeApiAsync methods (https://github.com/Azure/azure-mobile-apps/issues/808), it works fine, but sometimes we see this exception in our logs when we call the Serialize method:

System.InvalidOperationException: Collection was modified; enumeration operation may not execute.

![image](https://github.com/Azure/azure-mobile-apps/assets/9281751/e82fda31-3b81-4bb4-a439-5c17d8a36338)

## To Reproduce

It does not happen very often, so it is hard to reproduce, it seems to come from a concurrent access to the tableNameCache dictionary.

## Expected behavior

Not exception is thrown.

## What platforms?

* Client:
  * Version of dotnet being used to compile: net8
  * .NET Runtime Environment: WinAppSDK, iOS, Android
  * Library versions 6.1.0

## Additional context

The issue only happens when we call the Serialize method from our custom implementation of the InvokeApiAsync methods. Perhaps using a lock to query the tableNameCache dictionary in the DatasyncContractResolver.CreateProperties method would be enough to make this work fine.

```
lock (tableNameCache)
{
  if (!tableNameCache.ContainsKey(type) && !tableNameCache.Keys.Any(t => t.GetTypeInfo().IsAssignableFrom(typeInfo)))
  {
    return properties;
  }
}
```

**Comment by adrianhall** [comment]
Thanks for the report - I've been thinking of switching this all to ConcurrentDictionary types anyhow, so this is an impetus for it.

**Comment by asierpn** [comment]
But the ConcurrentDictionary's Keys collection is thread-safe? Another option is to make a copy of the Keys collection before querying it:

```
  if (!tableNameCache.ContainsKey(type) && !tableNameCache.Keys.ToList().Any(t => t.GetTypeInfo().IsAssignableFrom(typeInfo)))
  {
    return properties;
  }
```

**Comment by adrianhall** [comment]
Fixed in 6140ab3
